qemu/tests/cpu_device_hotpluggable_with_stress.py

```diff
@@ -0,0 +1,122 @@
+
+    os_type = params["os_type"]
+    login_timeout = params.get_numeric("login_timeout", 360)
+    stress_duration = params.get_numeric("stress_duration", 180)
+    verify_wait_timeout = params.get_numeric("verify_wait_timeout", 60)
+    vcpu_devices = params.objects("vcpu_devices")
+
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    session = vm.wait_for_login(timeout=login_timeout)
+    smp = vm.cpuinfo.smp
+    maxcpus = vm.cpuinfo.maxcpus
+
+    guest_cpu_ids = cpu_utils.get_guest_cpu_ids(session, os_type)
+    for vcpu_dev in vcpu_devices:
+        error_context.context("Hotplug vcpu device: %s" % vcpu_dev,
+                              logging.info)
+        vm.hotplug_vcpu_device(vcpu_dev)
+    if not utils_misc.wait_for(lambda: vm.get_cpu_count() == maxcpus,
+                               verify_wait_timeout):
+        test.fail("Actual number of guest CPUs is not equal to the "
+                  "expected.")
+
+    if os_type == "linux":
+        stress_args = params["stress_args"]
+        stress_tool = cpu_utils.VMStressBinding(vm, params,
+                                                stress_args=stress_args)
+        current_guest_cpu_ids = cpu_utils.get_guest_cpu_ids(session, os_type)
+        plugged_cpu_ids = list(current_guest_cpu_ids - guest_cpu_ids)
+        plugged_cpu_ids.sort()
+        for cpu_id in plugged_cpu_ids:
+            error_context.context("Run stress on vCPU(%d) inside guest."
+                                  % cpu_id, logging.info)
+            stress_tool.load_stress_tool(cpu_id)
+        error_context.context("Successfully launched stress sessions, execute "
+                              "stress test for %d seconds" % stress_duration,
+                              logging.info)
+        time.sleep(stress_duration)
+        if utils_package.package_install("sysstat", session):
+            error_context.context("Check usage of guest CPUs", logging.info)
+            mpstat_cmd = "mpstat 1 5 -P %s | cat" % ",".join(
+                map(str, plugged_cpu_ids))
+            mpstat_out = session.cmd_output(mpstat_cmd)
+            cpu_stat = dict(re.findall(r"Average:\s+(\d+)\s+(\d+\.\d+)",
+                                       mpstat_out, re.M))
+            for cpu_id in plugged_cpu_ids:
+                cpu_usage_rate = float(cpu_stat[str(cpu_id)])
+                if cpu_usage_rate < 50:
+                    test.error("Stress test on vCPU(%s) failed, usage rate: "
+                               "%.2f%%" % (cpu_id, cpu_usage_rate))
+                logging.info("Usage rate of vCPU(%s) is: %.2f%%", cpu_id,
+                             cpu_usage_rate)
+        for vcpu_dev in vcpu_devices:
+            error_context.context("Hotunplug vcpu device: %s" % vcpu_dev,
+                                  logging.info)
+            vm.hotunplug_vcpu_device(vcpu_dev)
+            # Drift the running stress task to other vCPUs
+            time.sleep(random.randint(5, 10))
```

Comment:
Why sleep random seconds?

Reply:
Actually, we don't know when the CPU task will drift successfully, but it should be fast. Here just simulate a random event.